File: web/src/engine/risk.js
// EV model, bet ramps and risk metrics.
// Mirrors src/bjtoolkit/{ev_model,ramp,risk,viability}.py.

import {
  HAND_VARIANCE, HAND_VARIANCE_TC_SLOPE, HILO_SLOPE,
} from "./constants.js";
import { counts, probAt } from "./frequency.js";
import { baseEdge } from "./rules.js";

export const evAtTc = (tc, rules, slope = null) =>
  baseEdge(rules) + (slope === null ? HILO_SLOPE.value : slope) * tc;

export const varianceAtTc = (tc, tcDependent = true) =>
  tcDependent
    ? HAND_VARIANCE.value + HAND_VARIANCE_TC_SLOPE.value * Math.max(0, tc)
    : HAND_VARIANCE.value;

export const breakevenTc = (rules) => -baseEdge(rules) / HILO_SLOPE.value;

const roundToUnit = (x, unit) => (unit > 0 ? Math.round(x / unit) * unit : x);

function rampShell(bets, satOut, warnings) {
  const played = {};
  for (const [t, b] of Object.entries(bets)) {
    if (!satOut.has(Number(t)) && b > 0) played[t] = b;
  }
  const values = Object.values(played);
  return {
    bets, satOut, warnings, played,
    bet: (tc) => (satOut.has(tc) ? 0 : bets[tc] || 0),
    spread: values.length ? Math.max(...values) / Math.min(...values) : 0,
    topBet: values.length ? Math.max(...values) : 0,
    bottomBet: values.length ? Math.min(...values) : 0,
  };
}

const wongSet = (dist, wongOutBelow) =>
  wongOutBelow === null || wongOutBelow === undefined
    ? new Set()
    : new Set(counts(dist).filter((t) => t < wongOutBelow));

export function kellyRamp(rules, dist, opts = {}) {
  const {
    bankroll, kellyFraction = 0.4, roundingUnit = null,
    maxSpread = null, wongOutBelow = null, tcDependentVariance = true,
  } = opts;
  const unit = roundingUnit === null ? rules.tableMin : roundingUnit;
  const satOut = wongSet(dist, wongOutBelow);
  const warnings = [];

  const raw = {};
  for (const tc of counts(dist)) {
    raw[tc] = Math.max(0, (kellyFraction * evAtTc(tc, rules) / varianceAtTc(tc, tcDependentVariance)) * bankroll);
  }

  let cap = rules.tableMax;
  if (maxSpread !== null) {
    cap = Math.min(cap, rules.tableMin * maxSpread);
    if (Object.values(raw).some((v) => v > rules.tableMin * maxSpread)) {
      warnings.push(`your cover limit of ${maxSpread}x is capping bets below what Kelly wants -- you are deliberately leaving EV on the table`);
    }
  }
  if (Object.values(raw).some((v) => v > rules.tableMax)) {
    warnings.push(`the ${rules.currency} ${rules.tableMax.toLocaleString()} table maximum is suppressing your top bets`);
  }

  const bets = {};
  for (const tc of counts(dist)) {
    const bet = Math.min(Math.max(raw[tc], rules.tableMin), cap);
    bets[tc] = Math.max(roundToUnit(bet, unit), rules.tableMin);
  }

  const advantage = counts(dist).filter((t) => evAtTc(t, rules) > 0 && !satOut.has(t));
  if (advantage.length) {
    const first = Math.min(...advantage);
    if (raw[first] < rules.tableMin) {
      const needed = (rules.tableMin * varianceAtTc(first, tcDependentVariance)) / (kellyFraction * evAtTc(first, rules));
      warnings.push(`bankroll too small for this table: at TC ${first >= 0 ? "+" : ""}${first} Kelly wants ${rules.currency} ${Math.round(raw[first]).toLocaleString()} but the minimum is ${rules.currency} ${rules.tableMin.toLocaleString()}. You would need about ${rules.currency} ${Math.round(needed).toLocaleString()} to bet this table properly at ${kellyFraction} Kelly.`);
    }
  }
  const ramp = rampShell(bets, satOut, warnings);
  if (ramp.spread && ramp.spread < 2) {
    warnings.push(`realised spread is only ${ramp.spread.toFixed(1)}x -- too flat to overcome the off-the-top house edge`);
  }
  return ramp;
}

// A Kelly-SHAPED ramp scaled to a chosen spread, independent of bankroll. This
// is what makes bankrollForRor a real solve rather than a fixed-point chase.
export function spreadRamp(rules, dist, opts = {}) {
  const {
    maxSpread, unit = null, wongOutBelow = null,
    roundingUnit = null, tcDependentVariance = true,
  } = opts;
  const u = unit === null ? rules.tableMin : unit;
  const rounding = roundingUnit === null ? u : roundingUnit;
  const satOut = wongSet(dist, wongOutBelow);
  const warnings = [];

  const shape = {};
  for (const tc of counts(dist)) {
    shape[tc] = Math.max(0, evAtTc(tc, rules)) / varianceAtTc(tc, tcDependentVariance);
  }
  const live = counts(dist).filter((t) => !satOut.has(t)).map((t) => shape[t]);
  const peak = live.length ? Math.max(...live) : 0;

  const bets = {};
  for (const tc of counts(dist)) {
    if (peak <= 0) { bets[tc] = u; continue; }
    const scaled = u + (maxSpread - 1) * u * (shape[tc] / peak);
    bets[tc] = Math.max(roundToUnit(scaled, rounding), u);
  }
  if (u * maxSpread > rules.tableMax) {
    warnings.push(`a ${maxSpread}x spread off a ${rules.currency} ${u.toLocaleString()} unit needs a top bet of ${rules.currency} ${(u * maxSpread).toLocaleString()}, above this table's maximum of ${rules.currency} ${rules.tableMax.toLocaleString()}`);
  }
  return rampShell(bets, satOut, warnings);
}

const phi = (x) => 0.5 * (1 + erf(x / Math.SQRT2));

function erf(x) {
  // Abramowitz & Stegun 7.1.26, good to about 1.5e-7 -- far tighter than the
  // model whose output it formats.
  const sign = x < 0 ? -1 : 1;
  x = Math.abs(x);
  const t = 1 / (1 + 0.3275911 * x);
  const y = 1 - ((((1.061405429 * t - 1.453152027) * t + 1.421413741) * t - 0.284496736) * t + 0.254829592) * t * Math.exp(-x * x);
  return sign * y;
}

export function evaluate(rules, dist, ramp, opts = {}) {
  const {
    bankroll = null, roundsPerHour = 70, hoursPlanned = 100,
    tcDependentVariance = true,
  } = opts;
  let wr = 0;
  let variance = 0;
  let playedP = 0;
  for (const tc of counts(dist)) {
    const p = probAt(dist, tc);
    const bet = ramp.bet(tc);
    if (bet <= 0) continue;
    playedP += p;
    wr += p * bet * evAtTc(tc, rules);
    variance += p * bet * bet * varianceAtTc(tc, tcDependentVariance);
  }
  const sd = variance > 0 ? Math.sqrt(variance) : 0;
  const warnings = [...ramp.warnings];
  if (wr <= 0) {
    warnings.push("this configuration is EV-NEGATIVE: you would lose money playing it. The ramp is too flat, the penetration too shallow, or the rules too poor to beat the house edge you pay on low counts.");
  }
  const n0Rounds = wr > 0 ? variance / (wr * wr) : Infinity;
  const expected = wr * roundsPerHour * hoursPlanned;
  const sdPlanned = sd * Math.sqrt(roundsPerHour * hoursPlanned);
  return {
    currency: rules.currency,
    fractionOfRoundsPlayed: playedP,
    winRatePerRound: wr,
    winRatePerPlayedHand: playedP > 0 ? wr / playedP : 0,
    variancePerRound: variance,
    sdPerRound: sd,
    winRatePerHour: wr * roundsPerHour,
    sdPerHour: sd * Math.sqrt(roundsPerHour),
    handsPlayedPerHour: playedP * roundsPerHour,
    n0Rounds,
    n0Hours: Number.isFinite(n0Rounds) ? n0Rounds / roundsPerHour : Infinity,
    bankroll,
    riskOfRuin: bankroll === null ? null : riskOfRuin(wr, variance, bankroll),
    hoursPlanned,
    expectedProfit: expected,
    sdOverPlanned: sdPlanned,
    probOfProfit: sdPlanned > 0 ? phi(expected / sdPlanned) : 0,
    warnings,
  };
}

export function riskOfRuin(winRatePerRound, variancePerRound, bankroll) {
  if (winRatePerRound <= 0) return 1;
  if (variancePerRound <= 0) return 0;
  return Math.exp((-2 * winRatePerRound * bankroll) / variancePerRound);
}

// Only meaningful for a ramp fixed in currency. With Kelly-sized bets, win rate
// scales as B and variance as B^2, so the ruin exponent is scale-invariant and
// this question has no answer -- use kellyForRor instead.
export function bankrollForRor(targetRor, winRatePerRound, variancePerRound) {
  if (!(targetRor > 0 && targetRor < 1)) throw new Error("targetRor must be between 0 and 1");
  if (winRatePerRound <= 0) return Infinity;
  return (-Math.log(targetRor) * variancePerRound) / (2 * winRatePerRound);
}

export const kellyForRor = (targetRor) => -2 / Math.log(targetRor);
export const rorForKelly = (f) => (f <= 0 ? 0 : Math.exp(-2 / f));

File: src/bjtoolkit/risk.py
"""Win rate, variance, N0, risk of ruin, and the bankroll solve.

A note on the solve, because the obvious version of it is wrong.

If your bets scale with your bankroll (which is what Kelly means), then
win_rate is proportional to B and variance is proportional to B^2, so the ruin
exponent 2*wr*B/var is scale-invariant: risk of ruin does not depend on your
bankroll at all. It depends only on what fraction of Kelly you bet. Trying to
"solve for the bankroll that gives 5% ruin" against Kelly-sized bets is a
fixed-point chase whose answer just tracks whatever bankroll you fed in.

So there are two honest questions, and this module answers both separately:

  bankroll_for_ror  -- given a FIXED ramp in currency, what bankroll do I need?
  kelly_for_ror     -- given proportional betting, what Kelly fraction may I use?

Wonging is handled by keeping everything per *round observed at the table*
rather than per round played. That way sitting out is just a bet of zero and the
arithmetic stays honest about the hours you spend not betting.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field

from .ev_model import ev_at_tc, variance_at_tc
from .frequency import TCDistribution
from .ramp import Ramp
from .rules import RuleSet


def _phi(x: float) -> float:
    """Standard normal CDF."""
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


@dataclass
class RiskReport:
    currency: str
    rounds_per_hour: int

    fraction_of_rounds_played: float
    win_rate_per_round: float
    """Per round OBSERVED at the table, counting sat-out rounds as zero."""
    win_rate_per_played_hand: float
    variance_per_round: float
    sd_per_round: float

    win_rate_per_hour: float
    sd_per_hour: float
    hands_played_per_hour: float

    n0_rounds: float
    n0_hours: float

    bankroll: float | None = None
    risk_of_ruin: float | None = None

    hours_planned: float = 0.0
    expected_profit: float = 0.0
    sd_over_planned: float = 0.0
    prob_of_profit: float = 0.0

    warnings: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            k: v for k, v in self.__dict__.items()
        }


def evaluate(
    rules: RuleSet,
    dist: TCDistribution,
    ramp: Ramp,
    *,
    bankroll: float | None = None,
    rounds_per_hour: int = 70,
    hours_planned: float = 100.0,
    tc_dependent_variance: bool = True,
) -> RiskReport:
    """Integrate the ramp against the true-count distribution."""
    wr = 0.0
    var = 0.0
    played_p = 0.0
    for tc in dist.counts():
        p = dist.p(tc)
        bet = ramp.bet(tc)
        if bet <= 0:
            continue
        played_p += p
        wr += p * bet * ev_at_tc(tc, rules)
        var += p * bet**2 * variance_at_tc(tc, tc_dependent=tc_dependent_variance)

    sd = math.sqrt(var) if var > 0 else 0.0
    warnings = list(ramp.warnings)

    if wr <= 0:
        warnings.append(
            "this configuration is EV-NEGATIVE: you would lose money playing it. "
            "The ramp is too flat, the penetration too shallow, or the rules too "
            "poor to beat the house edge you pay on low counts."
        )

    n0_rounds = var / wr**2 if wr > 0 else math.inf
    ror = None
    if bankroll is not None:
        ror = risk_of_ruin(wr, var, bankroll)

    expected = wr * rounds_per_hour * hours_planned
    sd_planned = sd * math.sqrt(rounds_per_hour * hours_planned)
    prob_profit = _phi(expected / sd_planned) if sd_planned > 0 else 0.0

    return RiskReport(
        currency=rules.currency,
        rounds_per_hour=rounds_per_hour,
        fraction_of_rounds_played=played_p,
        win_rate_per_round=wr,
        win_rate_per_played_hand=wr / played_p if played_p > 0 else 0.0,
        variance_per_round=var,
        sd_per_round=sd,
        win_rate_per_hour=wr * rounds_per_hour,
        sd_per_hour=sd * math.sqrt(rounds_per_hour),
        hands_played_per_hour=played_p * rounds_per_hour,
        n0_rounds=n0_rounds,
        n0_hours=n0_rounds / rounds_per_hour if math.isfinite(n0_rounds) else math.inf,
        bankroll=bankroll,
        risk_of_ruin=ror,
        hours_planned=hours_planned,
        expected_profit=expected,
        sd_over_planned=sd_planned,
        prob_of_profit=prob_profit,
        warnings=warnings,
    )


def risk_of_ruin(
    win_rate_per_round: float, variance_per_round: float, bankroll: float
) -> float:
    """Probability of losing the whole bankroll, playing forever.

    The standard exponential approximation. It assumes the ramp stays fixed in
    currency terms as the bankroll moves, which is true for a flat unit and false
    for strict Kelly resizing -- see the module docstring.
    """
    if win_rate_per_round <= 0:
        return 1.0
    if variance_per_round <= 0:
        return 0.0
    return math.exp(-2.0 * win_rate_per_round * bankroll / variance_per_round)


def bankroll_for_ror(
    target_ror: float, win_rate_per_round: float, variance_per_round: float
) -> float:
    """Bankroll needed to hold a FIXED ramp at a target risk of ruin.

    Only meaningful when the ramp is fixed in currency -- build it with
    `ramp.spread_ramp`, not `ramp.kelly_ramp`. Feeding this the output of a
    bankroll-sized ramp is the circularity described in the module docstring.
    """
    if not 0.0 < target_ror < 1.0:
        raise ValueError("target_ror must be strictly between 0 and 1")
    if win_rate_per_round <= 0:
        return math.inf
    return -math.log(target_ror) * variance_per_round / (2.0 * win_rate_per_round)


def kelly_for_ror(target_ror: float) -> float:
    """The Kelly fraction implied by a target risk of ruin.

    From RoR = exp(-2/f) for proportional betting: full Kelly gives ~13.5% ruin,
    half Kelly ~1.8%. This is the correct inverse when your bets scale with your
    bankroll, and it is independent of the bankroll's size.
    """
    if not 0.0 < target_ror < 1.0:
        raise ValueError("target_ror must be strictly between 0 and 1")
    return -2.0 / math.log(target_ror)


def ror_for_kelly(kelly_fraction: float) -> float:
    """Risk of ruin implied by proportional betting at this Kelly fraction."""
    if kelly_fraction <= 0:
        return 0.0
    return math.exp(-2.0 / kelly_fraction)
